### PR TITLE
Split monolithic ferc_to_sqlite ops into per-dataset pieces

### PR DESCRIPTION
**Pull request by rousik**
Presence of the monolithic `dbf2sqlite` and `xbrl2sqlite` ops that processed all datasets in  a serial loop seemed like a missed opportunity. This change rips out those monoliths and replaces them with dataset specific ops that are directly invoked from within the relevant graphs.

One thing that prevented efficient breakup into smaller ops was a fact that these ops relied on configuration fields like `clobber` and `batch_size`, and this could fan out quite drastically. To prevent this, I have replaced the op-specific config with `RuntimeSettings` resource that is defined once and for all and used by all the ops that need access to these parameters.

This change:
- decomposes `ferc_to_sqlite` graph into dataset specific operations that are generated by relevant extractor classes
- introduces `--dataset-only` flag that can further restrict the graph to single dataset operation (this is going to be used in CI for sharding and speeding up `ferc_to_sqlite`)
- drops support of `xbrl_enabled`/`dbf_enabled` flags in the job factory in favor of `op_selector` logic
- added tags to ops indicating which `dataset` and `data_format` is used in the op (so that they can be appropriately selected)

### CODE REVIEW

**Comment by rousik** [comment]
I ran this with the pytest refactor that runs `ferc_to_sqlite` as a standalone step on larger runner and got the following results:

  | 1 xbrl worker | 2 xbrl workers | 3 xbrl workers | 4 xbrl workers | 8 xbrl workers
-- | -- | -- | -- | -- | --
  | 49 mins | 43 mins | 41 mins | 40 mins
46 mins | 39 mins | 38 mins | 38 mins | ⛔
46 mins | 39 mins | ⛔ | ⛔ | ⛔
48 mins | 38 mins | ⛔ | ⛔ | ⛔

TL;DR - both dagster and xbrl worker parameters provide similar speed-ups, with xbrl workers providing slightly better value/speedups. Note that we can only run with 2 and 4 of each before we run larger runner out of memory. We could, therefore, run with 2 xbrl workers in `ci-integration` tests or explore running on even larger runners.

File: test/unit/extract/xbrl_test.py
Comment (jdangerx) [inline test/unit/extract/xbrl_test.py:104]: We're no longer mocking `PudlPaths()` here - is this comment still relevant? Seems like we're checking at test runtime for `--live-dbs`, which is fine I guess but also error-prone - we need to remember to bail out of any db-touching test when `live_dbs == True`.

File: test/unit/extract/xbrl_test.py
Comment (jdangerx) [inline test/unit/extract/xbrl_test.py:134]: I know I wrote these tests to actually hit disk, but that was probably a mistake. A unit test really shouldn't have to know if `--live-dbs` is set, because, well. It's a unit test.

I think that means we either need to: 
- mock out `PudlPaths().various_path_attributes` to return mocked Path objects, which we can then mock th `touch()`, `exists()` etc. API on
- mock out `PudlPaths` to always use tmp dir for output in unit tests, no matter what `--live-dbs` says - probably in `unit/conftest.py`

Do you have opinions or preference here? Personally I think the second one is harder to mess up in the future but I can definitely see an argument for "unit tests shouldn't be touching disk at all."

**Comment by zaneselvans** [comment]
Another option for splitting the sizeable FERC extraction work out from all of the other stuff without blowing up the resource use on an individual machine would be to split each of the FERC extractions into its own CI job. I think we only need to upload/download the FERC 1 DBs for the PUDL ETL to run. PUDL wouldn't need to be dependent on any of the other FERC jobs, and we could just upload and combine coverage reports from the non-FERC1 forms.

I think this could just be done by splitting the ferc2sqlite params out into separate settings files.

File: test/conftest.py
Comment (zaneselvans) [reply test/conftest.py:337]: I'm not 100% sure if/how this is related but, because we cache the inputs on GHA to avoid downloading them over and over again, they need to be in a fixed location on the runner and can't be in a randomly generated temporary directory (so the caching mechanism knows where to put the files, which happens before we start the tests)

IIRC there might also be some special cases where we *always* want to use a temporary directory e.g. to force a minimal test of the downloading facility even if we have cached inputs.

Comment (jdangerx) [reply test/conftest.py:337]: I think if we set `PUDL_INPUT` on GHA (which I think we do) then the caching at least should be unaffected by this change.

File: test/unit/extract/xbrl_test.py
Comment (rousik) [reply test/unit/extract/xbrl_test.py:134]: PudlPaths is instantiated from env variables so we could just temporarily set those and run whatever isolated tests we need to.

**Comment by rousik** [comment]
This is, in essence, what is done in the other PR that parallelizes ferc_to_sqlite. Sharing this into several workers run in parallel seems like yet another way to do what dagster should be doing for us our or the box, but I can see how this can be valuable thanks to machine/runner limits.

**Comment by rousik** [comment]
> Another option for splitting the sizeable FERC extraction work out from all of the other stuff without blowing up the resource use on an individual machine would be to split each of the FERC extractions into its own CI job. I think we only need to upload/download the FERC 1 DBs for the PUDL ETL to run. PUDL wouldn't need to be dependent on any of the other FERC jobs, and we could just upload and combine coverage reports from the non-FERC1 forms.
> 
> I think this could just be done by splitting the ferc2sqlite params out into separate settings files.  The FERC1 takes the most time by a significant margin. It could even be worthwhile to split out the XBRL and DBF extractions into 2 separate jobs (since they're totally independent conceptually, and take roughly the same amount of time to run). I think splitting the work out into separate CI jobs on separate runners would probably cut our ~45min FERC run down to more like 15min and would also cut the time we burn on larger runners almost in half.

This is basically putting yet another layer of doing the same thing in a different way on top of dagster parallelization; I can see the practical reason for it here (breaking out of the bounds of relatively small GitHub runners) but it's a bit of a dirty trick IMO. If this can save some money, then we could go for it. The other option could be to have `--only-dataset` or `--num-shards` flags that could subselect dagster assets/jobs or dataset settings programmatically, which would remove the need to maintain collection of configs for this.

**Comment by zaneselvans** [comment]
Yeah, I don't love the additional complexity, but it seems like it could be done with no change to the codebase, result in a significantly bigger speedup, and also save some paid CI minutes, so it seemed worth looking at.

@jdangerx Do you think there are any clear opportunities for speeding up the XBRL extraction jobs at a low level?  Right now it takes about the same amount of time to extract 20-30 years of DBF data as 2 years of XBRL data, so it seems like in production this issue is going to get worse over time.

File: src/pudl/resources.py
Comment (rousik) [reply src/pudl/resources.py:13]: Yeah, though I'm not sure whether these are compatible with `ConfigurableResource` base class. The way this is set up currently, we control this resource via existing command-line flags and perhaps that is good as-is. I will remove this comment.

File: test/conftest.py
Comment (rousik) [reply test/conftest.py:219]: Though this is probably irrelevant once we're running `ferc_to_sqlite` normally as part of `ci-integration` (see https://github.com/catalyst-cooperative/pudl/pull/2825) and not as part of integration tests in-process. At that point, we will likely deprecate these fixtures and assume we're always relying on `--live-dbs`?

File: test/conftest.py
Comment (rousik) [reply test/conftest.py:337]: Hmph, some tests were failing locally for me, I suppose `PUDL_INPUT` was not set then? I assume that we should always have this configured, and we don't expect to ever hit a scenario where this shouldn't be set, right? Maybe we should proactively abort here if the env variables are not set.

File: test/unit/extract/xbrl_test.py
Comment (rousik) [reply test/unit/extract/xbrl_test.py:104]: Yeah, the fact that we have the tests operating in these two distinct modes (read-only vs mutating) is a bit dangerous IMO and there's always a possibility that we don't do isolation well and things get mangled. I suppose that, ideally, we would rely on most of the db tests to be validations that will not modify whatever is in place.

**Comment by rousik** [comment]
> Another option for splitting the sizeable FERC extraction work out from all of the other stuff without blowing up the resource use on an individual machine would be to split each of the FERC extractions into its own CI job. I think we only need to upload/download the FERC 1 DBs for the PUDL ETL to run. PUDL wouldn't need to be dependent on any of the other FERC jobs, and we could just upload and combine coverage reports from the non-FERC1 forms.
> 
> I think this could just be done by splitting the ferc2sqlite params out into separate settings files. The FERC1 takes the most time by a significant margin. It could even be worthwhile to split out the XBRL and DBF extractions into 2 separate jobs (since they're totally independent conceptually, and take roughly the same amount of time to run). I think splitting the work out into separate CI jobs on separate runners would probably cut our ~45min FERC run down to more like 15min and would also cut the time we burn on larger runners almost in half.

I think this is a reasonable idea and I would like to pursue it as well. I'm thinking I could add `--dataset-only` and maybe `--format-only` flags to `ferc_to_sqlite` CLI that could filter down what ops are executed. I do think that having the work split into one op per dataset/format as is done here would pair very nicely with that functionality so I think we should still pursue this change.

For resource usage safety, I would also suggest setting default xbrl workers to 2 and rely primarily on the dagster workers for parallelism.

**Comment by rousik** [comment]
Implemented `--dataset-only=ferc1_xbrl` filter as a command-line flag that could be used in conjunction with the new `pytest`

**Comment by zaneselvans** [comment]
@rousik there was an error because `pudl` wasn't being imported in `conftest.py` so I added the import.

Now there's a real error related to splitting the FERC jobs up because the Datasette metadata to RST script that generates the `metadat.yaml` file that annotates our Datasette deployment expects all of the FERC outputs to be available in `$PUDL_OUTPUT` when it runs.

**Comment by rousik** [comment]
> @rousik there was an error because `pudl` wasn't being imported in `conftest.py` so I added the import.
> 
> Now there's a real error related to splitting the FERC jobs up because the Datasette metadata to RST script that generates the `metadat.yaml` file that annotates our Datasette deployment expects all of the FERC outputs to be available in `$PUDL_OUTPUT` when it runs.

@zaneselvans can you provide some additional context? Do we expect to run this datasette metadata job as part of `ci-integration`, or is that something that should be run as part of daily/nightly deployments? I suppose the question here is whether early coverage/error detection is worth the added cost (time/resources), or whether catching this on a nightly cadence (when we try to ship the datasette) is what we want.

**Comment by rousik** [comment]
> > @rousik there was an error because `pudl` wasn't being imported in `conftest.py` so I added the import.
> > Now there's a real error related to splitting the FERC jobs up because the Datasette metadata to RST script that generates the `metadat.yaml` file that annotates our Datasette deployment expects all of the FERC outputs to be available in `$PUDL_OUTPUT` when it runs.
> 
> @zaneselvans can you provide some additional context? Do we expect to run this datasette metadata job as part of `ci-integration`, or is that something that should be run as part of daily/nightly deployments? I suppose the question here is whether early coverage/error detection is worth the added cost (time/resources), or whether catching this on a nightly cadence (when we try to ship the datasette) is what we want.

It seems to me that we could adapt the datasette tests so that they're configuration aware and would only test for presence of files that should have been produced, i.e. skipping over datasets that are marked as disabled.

**Comment by zaneselvans** [comment]
I think we want to check whether we can generate the `metadata.yml` file that's required for Datasette prior to actually trying to do a deployment.  But also for testing purposes, it's probably fine if we just test it for `ferc1` and ignore the other forms, since if we break some underlying piece of infrastructure it'll probably often break all the forms, rather than just one of them.

**Comment by rousik** [comment]
Finally, I have figured out that `op_selection` can be used when converting graphs to jobs and that can be used to properly trim down the graph when `--dataset-only` flag is present. Simiarly, the flags to select only dbf/xbrl portions are now achieved by subselecting ops with `data_format=xbrl` tag attached to them. This simplifies the code somewhat.

**Comment by zaneselvans** [comment]
The main potential benefit I see here is being able to split off the FERC extractions into separate CI jobs that can run independently without blocking each other, and without needing a bigger runner, and then kicking off the main PUDL CI only after FERC1 has been extracted, so that only the PUDL CI runs on a big runner.  I’m not sure if this would be a good complexity for performance trade-off though.

I'm surprised that quadrupling the number of workers from 2 to 8 only results in a ~20% speedup. Does it fare better in local tests?

File: test/conftest.py
Comment (jdangerx) [reply test/conftest.py:337]: I think we should:
* remove this `gha and PUDL_INPUT not set` check, since that "should" never happen
* add a defensive check for PUDL_INPUT

File: test/unit/extract/xbrl_test.py
Comment (jdangerx) [inline test/unit/extract/xbrl_test.py:208]: Will this clobber the live SQLite DB if run with `--live-dbs`?

**Review by jdangerx — CHANGES_REQUESTED** [review]
Overall looks pretty good! Thanks for sticking with this long-standing PR for so long 😅 

I have a few clarifying questions to make sure the behavior is what we need before merging in.

File: docker/gcp_pudl_etl.sh
Comment (zaneselvans) [reply docker/gcp_pudl_etl.sh:137]: Yes, that was a mysterious side-effect that Jan noticed and that I'd come across before but never followed up on. We produce the `metadata.yml` file when we need it for the Datasette deployement separately.

File: test/conftest.py
Comment (zaneselvans) [reply test/conftest.py:219]: Are we going to rework the testing setup in the Makefile to run `ferc_to_sqlite` and `pudl_etl` before running the tests too?  That would be nice for the sake of consistency (doing the same thing in CI and locally) but would we need a different way of isolating the test outputs from live outputs? If the tests are always run with `--live-dbs` then should that option just be removed entirely?

File: src/pudl/ferc_to_sqlite/__init__.py
Comment (rousik) [reply src/pudl/ferc_to_sqlite/__init__.py:8]: I don't think we need those, removed.

File: src/pudl/resources.py
Comment (rousik) [reply src/pudl/resources.py:11]: yes, that's correct.

File: test/unit/extract/xbrl_test.py
Comment (rousik) [reply test/unit/extract/xbrl_test.py:208]: This method bails out early if `--live-dbs` is enabled. 

In the long run, we should ditch this flag and assume `--live-dbs` always, and ensure that we're not mucking with `PUDL_OUTPUT` in the tests at all.

I'm not aware if there's anything we could use to force read-only access on the PUDL_OUTPUT directory, so until then, we just have to rely on a good unit/integration test hygiene.

File: test/conftest.py
Comment (rousik) [reply test/conftest.py:219]: I'd say yes to removing `--live-dbs` functionality altogether, once we've isolated ferc_to_sqlite/etl from the unit/integration tests.

The obvious question is what is the semantics of running the integration tests locally - do we expect that the outputs are already in place, or do we want to re-run the data-generation (which could be time consuming).

In the latter case, even if we no longer have `--live-dbs`, we could set `PUDL_OUTPUT` in the `Makefile` itself to ensure that outputs that are generated for the sake of testing are isolated from the local outputs.

Happy to have some more conversation about this, but this seems to be out of scope for this specific PR so I would like to not block on that.

**Comment by rousik** [comment]
Reworked datasette tests after the recent changes there, to allow for presence of `ferc1_xbrl` only if `--live-dbs` is not enabled. This is not pretty, but I'm expecting we are going to ditch that functionality/duality and then we can fix these tests properly.

File: test/unit/extract/xbrl_test.py
Comment (jdangerx) [reply test/unit/extract/xbrl_test.py:208]: Ah, right, OK. Yeah, we should operate within the confines of our current `--live-dbs` world but scheme to overthrow the system and build a better world in the future.

File: src/pudl/resources.py
Comment (jdangerx) [reply src/pudl/resources.py]: Nag/reminder that you had meant to remove this comment.

**Review by jdangerx — APPROVED** [review]
Thanks for pushing this through @rousik ! Though there is still one merge conflict & a small typo re: a comment you had meant to remove.

There's definitely room for figuring out how we want to handle `--live-dbs` and that is also _way_ out of scope of this PR. If you have thoughts, @rousik, I'd love if you could write them up in a Discussion!

**Comment by rousik** [comment]
> Thanks for pushing this through @rousik ! Though there is still one merge conflict & a small typo re: a comment you had meant to remove.
> 
> There's definitely room for figuring out how we want to handle `--live-dbs` and that is also _way_ out of scope of this PR. If you have thoughts, @rousik, I'd love if you could write them up in a Discussion!

Merged `main` and resolved conflicts, letting the tests run overnight and fingers crossed, things should be good now.

File: src/pudl/resources.py
Comment (rousik) [reply src/pudl/resources.py]: This is pointing to an outdated file and the comment has been long gone. github is silly.

File: src/pudl/ferc_to_sqlite/cli.py
Comment (zaneselvans) [inline src/pudl/ferc_to_sqlite/cli.py:91]: Is the default here still supposed to be 2? If so the help message needs to be updated.

**Review by zaneselvans — COMMENTED** [review]
Hey @rousik note that we turned set up the merge queue last week, and the CI now only runs the integration tests in the merge queue.

File: src/pudl/ferc_to_sqlite/cli.py
Comment (jdangerx) [reply src/pudl/ferc_to_sqlite/cli.py]: Let's keep the default as 0, which is what it was, and then force concurrency of 2 in `conftest.py`. Also, we set `--workers 8` in `gcp_pudl_etl.sh`, so that is unaffected by this default.

**Comment by jdangerx** [comment]
Fixed merge conflict & changed the default back to 0. Though I think that will actually throw a `ValueError` per the [docs](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor), so I'm running the tests with a default of `None`.

**Pull request merged**
